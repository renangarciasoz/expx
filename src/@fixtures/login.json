{
  "username": "user@expx.fi",
  "password": "Br98PKe*js76QaF@1OdX"
}
