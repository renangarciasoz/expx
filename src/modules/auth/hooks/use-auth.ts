import { api } from "api/api-client";
import {
  recoverUserInformation,
  signInRequest,
  SignInRequestData,
} from "api/auth.api";
import { AxiosError } from "axios";
import { EXPIRATION_IN_HOURS, TOKEN_NAME } from "constants/auth.constants";
import { DASHBOARD, HOME } from "constants/urls.constants";
import { useRouter } from "next/router";
import { destroyCookie, parseCookies, setCookie } from "nookies";
import { useEffect } from "react";
import { useStore } from "store/auth.store";

export function useAuth() {
  const router = useRouter();
  const { user, error, setUser, setError, removeUser, ...store } = useStore();
  const isAuthenticated = !!user;

  useEffect(() => {
    const { [TOKEN_NAME]: token } = parseCookies();

    if (token) {
      recoverUserInformation().then((user) => {
        setUser(user);
      });
    }
  }, [setUser]);

  async function signIn({ email, password }: SignInRequestData) {
    setError("");

    try {
      const { token, ...user } = await signInRequest({
        email,
        password,
      });

      setCookie(undefined, TOKEN_NAME, token, {
        maxAge: 60 * 60 * EXPIRATION_IN_HOURS, // seconds * minutes * constant number in hours
      });

      api.defaults.headers.common["Authorization"] = `Bearer ${token}`;

      setUser(user);

      router.push(DASHBOARD);
    } catch (e) {
      removeUser();
      delete api.defaults.headers.common["Authorization"];
      setError((e as AxiosError).message);
    }
  }

  async function signOut() {
    destroyCookie(undefined, TOKEN_NAME);
    delete api.defaults.headers.common["Authorization"];
    removeUser();
    router.push(HOME);
  }

  return {
    ...store,
    isAuthenticated,
    user,
    error,
    signIn,
    setError,
    signOut,
  };
}
