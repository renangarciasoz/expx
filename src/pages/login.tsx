import {
  Box,
  Button,
  Container,
  Paper,
  TextField,
  Typography,
} from "@mui/material";
import { SignInRequestData } from "api/auth.api";
import { AuthContext } from "contexts/auth.ctx";
import type { NextPage } from "next";
import Head from "next/head";
import { useContext } from "react";
import { Controller, useForm } from "react-hook-form";

const Login: NextPage = () => {
  const { handleSubmit, control } = useForm<SignInRequestData>({
    defaultValues: {
      email: "",
      password: "",
    },
  });
  const { signIn } = useContext(AuthContext);

  function handleSignIn({ email, password }: SignInRequestData) {
    signIn({ email, password });
  }

  return (
    <Container>
      <Head>
        <title>Login</title>
      </Head>

      <Paper sx={{ p: 4 }}>
        <Typography variant="h1">Welcome</Typography>
        <Box
          component="form"
          onSubmit={handleSubmit(handleSignIn)}
          display="flex"
          flexDirection="column"
          m={2}
        >
          <Controller
            control={control}
            name="email"
            render={({ field }) => <TextField {...field} type="email" />}
          />

          <Controller
            control={control}
            name="password"
            render={({ field }) => <TextField {...field} type="password" />}
          />

          <Button type="submit">Sign in</Button>
        </Box>
      </Paper>
    </Container>
  );
};

export default Login;
