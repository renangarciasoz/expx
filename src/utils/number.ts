import { ONE_BILLION } from "constants/numbers.constants";

export const formatValue = (value: number, minimumFractionDigits = 2) =>
  Intl.NumberFormat("en", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits,
  }).format(value);

export const formatBillionValue = (
  value: string | number,
  minimumFractionDigits = 3
) => {
  return `${formatValue(Number(value) / ONE_BILLION, minimumFractionDigits)}b`;
};
