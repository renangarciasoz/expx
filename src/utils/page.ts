export const blockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "100vh";
  body!.style.overflow = "hidden";
};

export const unblockPageScroll = () => {
  const body = document?.querySelector("body");
  body!.style.height = "auto";
  body!.style.overflow = "auto";
};
