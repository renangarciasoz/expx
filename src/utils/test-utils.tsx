import { render, RenderOptions } from "@testing-library/react"; //jest-ignore
import { i18n } from "configs/i18next-tests.config";
import { ThemeProvider } from "contexts/theme.ctx";
import React, { ReactElement } from "react";
import { I18nextProvider } from "react-i18next";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
