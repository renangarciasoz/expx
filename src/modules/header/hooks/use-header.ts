import { POSITION_TOP } from "constants/header.constants";
import { useStore as useAuthStore } from "store/auth.store";
import { useStore } from "store/header.store";
import { blockPageScroll, unblockPageScroll } from "utils/page";

export const useHeader = () => {
  const { position, setFormAuthIsOpened, ...store } = useStore();
  const { setError } = useAuthStore();

  const headerIsAtTopPosition = position === POSITION_TOP;

  function openAuthForm() {
    blockPageScroll();
    setFormAuthIsOpened(true);
  }

  function closeAuthForm() {
    unblockPageScroll();
    setFormAuthIsOpened(false);
    setError(null);
  }

  return {
    openAuthForm,
    closeAuthForm,
    headerIsAtTopPosition,
    ...store,
  };
};
