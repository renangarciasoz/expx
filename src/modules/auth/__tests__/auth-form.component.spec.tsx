/**
 * @jest-environment jsdom
 */

import login from "@fixtures/login.json";
import * as useHeader from "modules/header/hooks/use-header";
import { act, cleanup, fireEvent, render, screen } from "utils/test-utils";
import { AuthForm } from "../auth-form.component";

jest.mock("next/router", () => ({
  ...jest.requireActual("next/router"),
  useRouter: () => ({
    push: jest.fn(),
  }),
}));

describe("Auth Form", () => {
  describe("when renders form", () => {
    it("renders correctly", () => {
      const { container } = render(<AuthForm />);
      expect(container).toMatchSnapshot();
    });

    describe("when header is at bottom position", () => {
      beforeAll(() => {
        jest.spyOn(useHeader, "useHeader").mockImplementation(() => ({
          ...jest.requireActual("modules/header/hooks/use-header"),
          headerIsAtTopPosition: false,
        }));
      });

      it("auth wrapper should have 80px top margin", () => {
        const { getByTestId } = render(<AuthForm />);
        const formWrapperStyles = getComputedStyle(getByTestId("auth-wrapper"));

        expect(formWrapperStyles.marginTop).toBe("80px");
      });
    });

    describe("when user does fill the form and submit", () => {
      describe("with wrong credentials", () => {
        beforeEach(async () => {
          const { getByTestId } = render(<AuthForm />);
          const emailInput = getByTestId("email-input");
          const passwordInput = getByTestId("password-input");
          const loginForm = getByTestId("auth-form");

          await act(async () => {
            await fireEvent.change(emailInput, {
              target: { value: "wrong@wrong.com" },
            });
            await fireEvent.change(passwordInput, {
              target: { value: "wrongpassword" },
            });
            await fireEvent.submit(loginForm);
          });
        });

        afterEach(cleanup);

        it("renders error message", () => {
          expect(screen.getByTestId("login-error-msg")).toBeInTheDocument();
        });
      });

      describe("with wrong credentials", () => {
        beforeEach(async () => {
          jest.spyOn(useHeader, "useHeader").mockImplementation(() => ({
            ...jest.requireActual("modules/header/hooks/use-header"),
            closeAuthForm: jest.fn(),
          }));

          const { getByTestId } = render(<AuthForm />);
          const emailInput = getByTestId("email-input");
          const passwordInput = getByTestId("password-input");
          const loginForm = getByTestId("auth-form");

          await act(async () => {
            await fireEvent.change(emailInput, {
              target: { value: login.username },
            });

            await fireEvent.change(passwordInput, {
              target: { value: login.password },
            });

            await fireEvent.submit(loginForm);
          });
        });

        afterEach(cleanup);

        it("not renders error message", () => {
          expect(
            screen.queryByTestId("login-error-msg")
          ).not.toBeInTheDocument();
        });
      });
    });
  });
});
