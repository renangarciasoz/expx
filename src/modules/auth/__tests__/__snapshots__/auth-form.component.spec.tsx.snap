// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Form when renders form renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-bdvt7g"
  >
    <div
      class="MuiBox-root css-bb7pjn"
      data-testid="auth-wrapper"
    >
      <h1
        class="MuiTypography-root MuiTypography-h1 MuiTypography-alignCenter css-ibvdv0-MuiTypography-root"
      >
        Login
      </h1>
      <form
        class="MuiBox-root css-1w0q4w9"
        data-testid="auth-form"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-13ro6dx-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-40493w-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-1"
            id="mui-1-label"
          >
            E-mail
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-mhnoc2-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="email-input"
              id="mui-1"
              name="email"
              type="email"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  E-mail
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-mf9yet-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-40493w-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="mui-2"
            id="mui-2-label"
          >
            Password
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-10gdzrp-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="password-input"
              id="mui-2"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root  css-yk7koz-MuiButtonBase-root-MuiButton-root"
          data-testid="login-submit-btn"
          tabindex="0"
          type="submit"
        >
          Sign in
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</div>
`;
